### PR TITLE
part 2 of the distroless/static move

### PR DESCRIPTION
**Pull request by mythi**
Closes: #516

### CODE REVIEW

**Comment by mythi** [comment]
TODO

- [ ]  intel-vpu-plugin
- [ ]  copy toybox/musl license

File: build/docker/intel-fpga-initcontainer.Dockerfile [hunk 1]
```diff
@@ -1,73 +1,62 @@
...
 
-RUN echo -e "#!/bin/sh\n\
-rsync -a --delete $SRC_DIR/ $DST_DIR\n\
-mkdir -p /etc/containers/oci/hooks.d\n\
-ln -sf $HOOK_CONF_DST /etc/containers/oci/hooks.d/$HOOK_CONF\n\
-rm $DST_DIR/deploy.sh\n\
-">> $SRC_DIR/deploy.sh && chmod +x $SRC_DIR/deploy.sh
+ARG TOYBOX_VERSION="0.8.4"
+RUN apt update && apt -y install musl musl-tools musl-dev
+RUN curl -SL https://github.com/landley/toybox/archive/refs/tags/$TOYBOX_VERSION.tar.gz | tar xz \
```
Comment (bart0sh) [inline build/docker/intel-fpga-initcontainer.Dockerfile:55]: Would it make sense to have toybox in a base container that we can use for initcontainer builds?

**Comment by mythi** [comment]
@bart0sh @rojkov this is now ready

File: build/docker/intel-fpga-initcontainer.Dockerfile [hunk 2]
```diff
@@ -1,73 +1,64 @@
...
-    cd $DIR && \
-    install -D ${DIR}/LICENSE /install_root/usr/local/share/package-licenses/intel-device-plugins-for-kubernetes/LICENSE && \
-    scripts/copy-modules-licenses.sh ./cmd/fpga_crihook /install_root/usr/local/share/ && \
-    scripts/copy-modules-licenses.sh ./cmd/fpga_tool /install_root/usr/local/share/
-
-# Minimal result image
-FROM scratch as final
-COPY --from=builder /install_root /
+RUN cd cmd/fpga_crihook && GO111MODULE=${GO111MODULE} CGO_ENABLED=0 go install "${BUILDFLAGS}" && \
+    cd ../fpga_tool && GO111MODULE=${GO111MODULE} CGO_ENABLED=0 go install "${BUILDFLAGS}" && \
```
Comment (bart0sh) [inline build/docker/intel-fpga-initcontainer.Dockerfile:34]: Do we still need GO11MODULE and CGO_ENABLED everywhere?

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: @bart0sh CI builds with `make vendor` so the image build uses `GO111MODULE=off`:
https://github.com/intel/intel-device-plugins-for-kubernetes/blob/a207bfe4ed3dfcd7cb68e024d5d4fe5510ab8bac/build/docker/build-image.sh#L27-L30

CGO must be disabled to be able to build static Go binaries without c-runtime.

Comment (rojkov) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: nit: perhaps CGO could be enabled explicitly for VPU then to future-proof the defaults.

**Review by rojkov — DISMISSED** [review]
LGTM

File: build/docker/intel-fpga-initcontainer.Dockerfile (hunk 2 above)
Comment (bart0sh) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: And why do we need to build static go binaries if we're running them in defined environment?

BTW, I don't see any difference in the fpga_tool size with and without CGO:
```
$ rm -f ~/go/bin/fpga_tool && CGO_ENABLED=1 go install "-ldflags=-w -s" && ls -la ~/go/bin/fpga_tool
-rwxrwxr-x 1 ed ed 2420736 touko 31 12:52 /home/ed/go/bin/fpga_tool

$ rm -f ~/go/bin/fpga_tool && CGO_ENABLED=0 go install "-ldflags=-w -s" && ls -la ~/go/bin/fpga_tool
-rwxrwxr-x 1 ed ed 2420736 touko 31 12:53 /home/ed/go/bin/fpga_tool
```

Comment (bart0sh) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: >  CI builds with make vendor so the image build uses GO111MODULE=off

Should we get rid of it and have consistent builds in CI and default environments?

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: > And why do we need to build static go binaries if we're running them in defined environment?

right, it's not required for the hooks that have the runtime libs available. with `CGO_ENABLED=0`, the NFD (GPU/SGX) hooks work also when nfd-worker is run static image.

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: > nit: perhaps CGO could be enabled explicitly for VPU then to future-proof the defaults.

make sense

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: > Should we get rid of it and have consistent builds in CI and default environments?

this is basically the topic of #565. we'd need to find the way to get go mod cache re-used by the container image builds.

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: > nit: perhaps CGO could be enabled explicitly for VPU then to future-proof the defaults.

enabled

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: @bart0sh do you have opens still?

Comment (bart0sh) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: @mythi yes, still wondering why do we need to repeat CGO_ENABLED=0 everywhere.

Comment (mythi) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: @bart0sh we want static binaries. I could move it to an `ENV` override too but it's almost the same

Comment (bart0sh) [reply build/docker/intel-fpga-initcontainer.Dockerfile:34]: we've just discussed it and I'm now convinced. Resolved.

**Review by bart0sh — APPROVED** [review]
/lgtm

**Pull request merged**